File: Python/DiffieHellman.py
import random

# Функция для вычисления степени по модулю
def powmod(a, x, p):
    res = 1
    for i in range(1, x + 1):
        res = res * a
        res = res % p
    return res

# Функция шифрования
def encoder(message, key):
    encrypted = ""
    for c in message:
        encrypted += chr(ord(c) ^ key)
    return encrypted

# Функция расшифровки
def decoder(message, key):
    decrypted = ""
    for c in message:
        decrypted += chr(ord(c) ^ key)
    return decrypted

# Решето Эратосфена для нахождения простых чисел до n
def sieveEratosthenes(n):
    is_prime = [True] * (n + 1)
    primes = []
    
    is_prime[0] = is_prime[1] = False  # 0 и 1 не являются простыми
    
    for i in range(2, int(n ** 0.5) + 1):
        if is_prime[i]:
            for j in range(i * i, n + 1, i):
                is_prime[j] = False  # отмечаем все кратные i как составные
    
    for i in range(2, n + 1):
        if is_prime[i]:
            primes.append(i)
    
    return primes

n1 = 100
n2 = 50
primes = sieveEratosthenes(n1)

# Генератор случайных чисел
random.seed()
dist1 = random.randint(0, len(primes) - 1)

p = primes[dist1]
g = 3

print(f"g = {g}; p = {p}")

a = random.randint(1, n2)
b = random.randint(1, n2)

print(f"Private key a = {a}")
print(f"Private key b = {b}")

A = powmod(g, a, p)
B = powmod(g, b, p)

print(f"Public key A = {A}")
print(f"Public key B = {B}")

sA = powmod(B, a, p)
sB = powmod(A, b, p)

print(f"Shared secret key sA = {sA}")
print(f"Shared secret key sB = {sB}")

message = "Hello World!"
encrypted = encoder(message, sA)
print(f"Message: {message}")
print(f"Encrypted data with sA: {encrypted}")

decrypted = decoder(encrypted, sB)
print(f"Decrypted data with sB: {decrypted}")

File: C++/DiffieHellman.cpp
#include <iostream>
#include <random>
using namespace std;

int powmod(int a, int x, int p) {
    int res = 1;
    for (int i = 1; i <= x; i++) {
        res = res * a;
        res = res % p;
    }
    return res;
}

string encoder(string message, int key) {
    string encrypted = "";
    for (char c : message) {
        encrypted += c ^ key;
    }
    return encrypted;
}

string decoder(string message, int key) {
    string decrypted = "";
    for (char c : message) {
        decrypted += c ^ key;
    }
    return decrypted;
}

vector<int> sieveEratosthenes(int n) {
    vector<bool> is_prime(n + 1, true);
    vector<int> primes;
    
    is_prime[0] = is_prime[1] = false; // 0 и 1 не являются простыми
    
    for (int i = 2; i * i <= n; i++) {
        if (is_prime[i]) {
            for (int j = i * i; j <= n; j += i) {
                is_prime[j] = false; // отмечаем все кратные i как составные
            }
        }
    }
    
    for (int i = 2; i <= n; i++) {
        if (is_prime[i]) {
            primes.push_back(i);
        }
    }
    
    return primes;
}

int main() {

    int n1 = 100;
    int n2 = 50;
    vector<int> primes = sieveEratosthenes(n1);

    random_device rd;
    mt19937 gen(rd());
    uniform_int_distribution<> dist1(0, primes.size() - 1);

    int p = primes[dist1(gen)];
    int g = 3;

    cout << "g = " << g << "; " << "p = " << p << endl;

    uniform_int_distribution<> dist2(1, n2);

    int a = dist2(gen);
    int b = dist2(gen);

    cout << "Private key a = " << a << endl;
    cout << "Private key b = " << b << endl;

    int A = powmod(g, a, p);
    int B = powmod(g, b, p);

    cout << "Public key A = " << A << endl;
    cout << "Public key B = " << B << endl;

    int sA = powmod(B, a, p);
    int sB = powmod(A, b, p);

    cout << "Shared secret key sA = " << sA << endl;
    cout << "Shared secret key sB = " << sB << endl;

    string message = "Hello World!";
    string encrypted = encoder(message, sA);
    cout << "Message: "  << message << endl;
    cout << "Encrypted data with sA: " << encrypted << endl;

    string decrypted = decoder(encrypted, sB);
    cout << "Decrypted data with sB: " << decrypted << endl;
}



